cpp/src/parquet/metadata3_benchmark.cc

```diff
@@ -0,0 +1,611 @@
+  AppendUleb(ext.size(), &thrift);
+  thrift += ext;
+  thrift += '\x00';  // add the stop field
+  return thrift;
+}
+
+void EncodeFlatbuf(benchmark::State& state, const Footer& footer) {
+  auto md = footer.md;
+  for (auto _ : state) {
+    auto ser = ToFlatBuffer(&md);
+    benchmark::DoNotOptimize(ser);
+  }
+}
+
+void ParseAndVerifyFlatbuf(benchmark::State& state, const Footer& footer) {
+  for (auto _ : state) {
+    flatbuffers::Verifier v(reinterpret_cast<const uint8_t*>(footer.flatbuf.data()),
+                            footer.flatbuf.size());
+    auto fmd = format3::GetFileMetaData(footer.flatbuf.data());
+    DCHECK_EQ(fmd->num_rows(), footer.md.num_rows);
+    bool ok = fmd->Verify(v);
+    DCHECK(ok);
+  }
+}
+
+void ParseWithExtension(benchmark::State& state, const Footer& footer) {
+  auto with_ext = AppendExtension(footer.thrift, footer.flatbuf);
+
+  for (auto _ : state) {
+    auto md = DeserializeThrift(with_ext);
+  }
+}
+
+void Analyze(std::string_view name, const format::FileMetaData& md) {
+  std::cerr << "Analyzing: " << name << "\n";
+  std::vector<int> sizes;
+  int num_cols = md.schema.size() - 1;
+  size_t stats_bytes = 0, kv_bytes = 0;
+  for (auto& kv : md.key_value_metadata) kv_bytes += kv.key.size() + kv.value.size();
+  for (int i = 0; i < num_cols; ++i) {
+    for (auto& rg : md.row_groups) {
+      auto& cc = rg.columns[i];
+      if (!cc.__isset.meta_data) continue;
+      auto& cmd = cc.meta_data;
+      auto& s = cmd.statistics;
+      stats_bytes += s.max_value.size() + s.min_value.size();
+      for (auto& kv : cmd.key_value_metadata) kv_bytes += kv.key.size() + kv.value.size();
+    }
+  }
+  std::cerr << "num-rgs=" << md.row_groups.size() << " num-cols=" << num_cols
+            << " stats_bytes=" << stats_bytes << " kv_bytes=" << kv_bytes << "\n";
+}
+
+}  // namespace
+}  // namespace parquet
+
+int main(int argc, char** argv) {
+  ::benchmark::Initialize(&argc, argv);
+  std::vector<parquet::Footer> footers;
+  for (int i = 1; i < argc; ++i) footers.push_back(parquet::Footer::Make(argv[i]));
+  struct {
+    std::string name;
+    void (*fn)(benchmark::State&, const parquet::Footer&);
+  } benchmarks[] = {
+      {"Parse", parquet::Parse},
+      {"ParseWithExtension", parquet::ParseWithExtension},
+      {"EncodeFlatbuf", parquet::EncodeFlatbuf},
+      {"ParseAndVerifyFlatbuf", parquet::ParseAndVerifyFlatbuf},
+  };
+  for (auto&& footer : footers) {
+    for (auto&& [n, fn] : benchmarks) {
+      ::benchmark::RegisterBenchmark(n + "/" + footer.name, fn, footer)
+          ->Unit(benchmark::kMillisecond);
+    }
+  }
+
+  char key[1024];
+  char val[1024];
+  for (size_t i = 0; i < footers.size(); ++i) {
+    auto&& f = footers[i];
+    snprintf(key, sizeof(key), "%lu/%s", i, basename(f.name.c_str()));
+    snprintf(val, sizeof(val), "num-rgs=%lu num-cols=%lu thrift=%lu flatbuf=%lu",
+             f.md.row_groups.size(), f.md.schema.size(), f.thrift.size(),
+             f.flatbuf.size());
```

Comment:
I found the results a bit hard to read without comma separators. Maybe format with commas:
```
snprintf(val, sizeof(val), "num-rgs=%'lu num-cols=%'lu thrift=%'lu flatbuf=%'lu",
             f.md.row_groups.size(), f.md.schema.size(), f.thrift.size(),
             f.flatbuf.size());
 ```
             
Along with
```
include<locale>

and

setlocale(LC_ALL, ""); // to enable comma formatting
```      

Gives output like
```
0/benchmark_float32_1000cols.footer: num-rgs=1 num-cols=1,000 thrift=109,811 flatbuf=96,320
```

Reply:
Also, there is an off-by-one error here. The num-cols output should be:
`f.md.schema.size()-1`

Reply:
Thank you @corwinjoy! I fixed the off by one. Instead of locale I pushed a change that outputs SI units. I rewrote history to make the fix part of the first commit.